pypesto/sampling/pymc3.py

```diff
@@ -112,11 +105,221 @@ def get_samples(self) -> McmcPtResult:
+        else:
+            return 'free_parameters'
+    else:
+        return 'theta'
+
+
+def pymc3_to_arviz(model: pm.Model, trace: pm.backends.base.MultiTrace, problem: Problem = None, vectorize: bool = False):
+    kwargs = {}
+    if vectorize:
+        if problem is None:
+            raise ValueError('if vectorize is True, ' \
+                             'a problem must be given')
+        x_free_names = [problem.x_names[i] for i in problem.x_free_indices]
+        kwargs['coords'] = {"free_parameter": x_free_names}
+        kwargs['dims'] = {pymc3_vector_parname(problem): ["free_parameter"]}
+
+    return az.from_pymc3(
+        trace=trace,
+        model=model,
+        log_likelihood=False,  # NB at the moment we do not separate prior
+                               # from posterior, thus computing likelihood
+                               # from pymc3 model is meaningless
+        **kwargs
+    )
+
+
+def BetterUniform(name, *, lower, upper, lerp='convex', **kwargs):
+    """
+    A uniform bounded random variable whose behaviour near the boundary of
+    the domain is better than the native `pymc3.Uniform`.
+
+    The problem with `pymc3.Uniform` is that the inverse transform formula
+    fails in floating point arithmetic when the lower and the upper bound differ
+    largely in magnitude, resulting in a value outside the original
+    interval. This leads the log-posterior for the original distribution to
+    become `-inf`. By using a better formula this can be avoided.
+
+    While this is a sufficient fix, there is another opportunity for
+    simplification: avoiding the computation of `log(ub - lb)` in the
+    log-posterior (it appears both in the original log-posterior and in the
+    transform jacobian, so it can be simplified away).
+    This simplification is imposed by using instead of `Uniform` a `Flat` prior
+    (which has log-posterior 0) and removing the term `log(ub - lb)` from
+    the interval transformation jacobian.
+    """
+    if 'shape' not in kwargs.keys():
+        if 'testval' in kwargs.keys():
+            shape = np.broadcast(lower, upper, kwargs['testval']).shape
+        else:
+            shape = np.broadcast(lower, upper).shape
+        kwargs['shape'] = shape
+    scalar = (len(kwargs['shape']) == 0)
+
+    if 'transform' in kwargs.keys():
+        raise Exception('if specifying a custom transform, ' \
+                        'please use pymc3.Uniform')
+
+    BoundedFlat = pm.Bound(pm.Flat, lower=lower, upper=upper)
+
+    transform = BetterInterval(lower, upper, lerp, scalar=scalar)
+
+    return BoundedFlat(name, transform=transform, **kwargs)
+
+    # In the case we start from pm.Uniform,
+    # we need to comment the jacobian out of BetterInterval
+    # transform = BetterInterval(lower, upper, lerp, scalar=scalar)
+    # return pm.Uniform(name, lower=lower, upper=upper, transform=transform, **kwargs)
+
+
+class BetterInterval(pm.distributions.transforms.Interval):
```

Comment:
add doc

Reply:
consider adding unittests for this class that check that intended behavior is achieved.

Reply:
in particular, it must be ensured that the sampling distribution is not changed.